src/common/classes/array.h

```diff
@@ -425,6 +425,16 @@ class Array : protected Storage
 		return data;
 	}
 
+	// prepare array to be used as a buffer of capacity bytes aligned on given alignment
+	T* getAlignedBuffer(const size_type capacityL, const size_type alignL)
+	{
+		static_assert(sizeof(T) == 1, "sizeof(T) != 1");
+
+		ensureCapacity(capacityL + alignL, false);
```

Comment:
I think that `- ALLOC_ALIGNMENT` can be added here to reduce memory overhead a little.

Reply:
AFAIU your code add some unused space at the beginning and the end of allocated buffer to get right alignment. Because allocated raw buffer already aligned, the difference between raw buffer pointer and returning buffer pointer is guaranteed to be smaller than `alignL` at least by ALLOC_ALIGNMENT (actually ALLOC_ALIGNMENT + 1) bytes, so unused space at the end can be safely reduced.

Reply:
> AFAIU your code add some unused space at the beginning and the end of allocated buffer to get right alignment. Because allocated raw buffer already aligned, the difference between raw buffer pointer and returning buffer pointer is guaranteed to be smaller than `alignL` at least by ALLOC_ALIGNMENT (actually ALLOC_ALIGNMENT + 1) bytes, so unused space at the end can be safely reduced.

Formally you right, but how mush you going to save ? 16 bytes of few KBs ?

Reply:
Do you really think that every page buffer in page cache is allocated separately ?

Reply:
> Do you really think that every page buffer in page cache is allocated separately ?

Probably I'm wrong but yes, I do.

Reply:
We're talking about generic code in common, so I support the idea of optimize it for general use case.

Reply:
Ok. In any case the difference can be found on https://firebirdtest.com/oltp-emul/ if someone add this branch there.

Reply:
> We're talking about generic code in common, so I support the idea of optimize it for general use case.

Sorry, I see any real benefits here but code complication.
Feel free to add such optimization outside of this PR.